File: 239. 滑动窗口最大值.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function(nums, k) {
    const p = new PriorityQueue({
        compare: (a, b)=>{
            return a.value < b.value;
        }
    });
    
    const out = [];

    for (let i = 0; i < k; i++) {
        p.enqueue({
            value: nums[i],
            index: i
        });
    }
    out.push(p.front().value);
    for (let i = k; i < nums.length; i++) {
        p.enqueue({
            value: nums[i],
            index: i
        });
        while (p.front().index <= i - k) {
            p.dequeue();
        }
        out.push(p.front().value);
    }
    return out;
};

File: 239. 滑动窗口最大值.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        priority_queue<pair<int, int>> p;

        vector<int> out;
        for (int i = 0; i < k; i++) {
            p.emplace(nums[i], i);
        }
        out.push_back(p.top().first);
        for (int i = k; i < nums.size(); i++) {
            p.emplace(nums[i], i);
            while (p.top().second <= i - k) {
                p.pop();
            }
            out.push_back(p.top().first);
        }
        return out;
    }
};